### ISSUE TITLE
azurerm_mssql_server_extended_auditing_policy does not correctly set up access to storage account

### ISSUE THREAD

**Issue by eugeneromero** [terraform-providers/terraform-provider-azurerm#9170]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

`Terraform v0.13.5`
`azurerm v2.34.0`

### Affected Resource(s)

* `azurerm_mssql_server`
* `azurerm_mssql_server_extended_auditing_policy`

### Terraform Configuration Files

```hcl
resource "azurerm_mssql_server" "sqlDbServer" {
  name                          = "sql"
  resource_group_name           = "rg"
  location                      = "norwayeast"
  version                       = "12.0"
  administrator_login           = "sqladmin"
  administrator_login_password  = "password"
  public_network_access_enabled = false

  identity {
    type = "SystemAssigned"
  }
  tags = {var="foo"}
}

resource "azurerm_storage_account" "sqlAuditingStorage" {
  name                     = "storage"
  resource_group_name      = "rg"
  location                 = "norwayeast"
  account_tier             = "Standard"
  account_replication_type = "LRS"
}

resource "azurerm_mssql_server_extended_auditing_policy" "sqlDbServerAuditingPolicy" {
  server_id                               = azurerm_mssql_server.sqlDbServer.id
  storage_endpoint                        = azurerm_storage_account.sqlAuditingStorage.primary_blob_endpoint
  storage_account_access_key              = azurerm_storage_account.sqlAuditingStorage.primary_access_key
  storage_account_access_key_is_secondary = false
  retention_in_days                       = 7
  tags = {var="foo"}
}
```

### Expected Behavior
On first run, Terraform should create the SQL server, storage account, and extended auditing policy. On subsequent runs, Terraform/Azure should not throw an error.

### Actual Behavior
On first run, Terraform creates the resources correctly. Access is working correctly, because I can go to the storage account and see the SQL auditing logs. However, if I change anything in the SQL or auditing policy's resource definitions (say, a tag) so that Terraform must update it, Terraform errors out.

### Steps to Reproduce
Set up the above resources. Apply with TF. Then, make any change to the resources (a tag is fine). On second apply, Terraform throws this error:

```
Code="BlobAuditingInsufficientStorageAccountPermissions" Message="Insufficient read or write permissions on storage account 'storage'. Add permissions to the server Identity to the storage account."
```

### Important Factoids
1. At first, based on the error, I thought this might have to do with the SQL server's Service Principal needing access to the storage account. So I added an additional `azuread` resource to give the SP "Contributor" role assignment in the Storage account. At this point, the error changes:

```
Error: waiting for creation of SQL Server "sql" Blob Auditing Policies(Resource Group "rg"): Code="InternalServerError" Message="An unexpected error occured while processing the request. Tracking ID: [...]"
```
After some more testing, I realized that having this assignment does not make any difference in the final result, so I removed it.

2. I decided to run `terraform apply` with the debug flag, and this caught my attention:

```
"type": "azurerm_mssql_server",
"name": "sqlDbServer",
"provider": "provider[\"registry.terraform.io/hashicorp/azurerm\"]",
"instances": [
  {
    "attributes": {
      [...]
      "extended_auditing_policy": [
        {
          "retention_in_days": 7,
          "storage_account_access_key": "",             <-------- THIS IS EMPTY
          "storage_account_access_key_is_secondary": false,
          "storage_endpoint": [REDACTED]              <-------- THIS IS POPULATED
        }
      ],

-------------------------

"type": "azurerm_mssql_server_extended_auditing_policy",
"name": "sqlDbServerAuditingPolicy",
"provider": "provider[\"registry.terraform.io/hashicorp/azurerm\"]",
"instances": [
  {
    "attributes": {
      "retention_in_days": 7,
      "server_id": "[ID]",
      "storage_account_access_key": [REDACTED],    <-------- THIS IS POPULATED
      "storage_account_access_key_is_secondary": false,
      "storage_endpoint": [REDACTED],                        <-------- THIS IS POPULATED
      "timeouts": null
    },
```
so I started wondering if the issue could be related to the missing `storage_account_access_key` value. Even though I do not have an `extended_auditing_policy` in the sql server definition, I found it odd that the endpoint value _was_ populated. So, to test, I removed the `azurerm_mssql_server_extended_auditing_policy` resource and moved it into an `extended_auditing_policy` block inside the `azurerm_mssql_server`. Sure enough, this solves the issue. So, I wonder if Terraform is only trying to use the access key from the `server` definition, and not from the `policy` definition?

Of course, using it like this, I get a deprecation warning:
```
Warning: "extended_auditing_policy": [DEPRECATED] the `extended_auditing_policy` block has been moved to `azurerm_mssql_server_extended_auditing_policy` and `azurerm_mssql_database_extended_auditing_policy`. This block will be removed in version 3.0 of the provider.
```
However, I think this bug should be fixed before removing the old definition, because it currently breaks CD pipelines with the new resource.

Thanks for your attention to this bug report!

**Comment by yupwei68** [comment]
Hi @eugeneromero , thanks for opening this issue. Please refer to the document https://docs.microsoft.com/en-us/azure/azure-sql/database/audit-write-storage-account-behind-vnet-firewall. 
For the first problem, auditing requires extra permission to storage account behind firewall.
For the second problem, but the permission is "Storage Blob Data Contributor" instead of "Contributor" role in the scope of storage account.
An example would be:
```
resource "azurerm_role_assignment" "test" {
  scope                = azurerm_storage_account.test.id
  role_definition_name = "Storage Blob Data Contributor"
  principal_id         = azurerm_mssql_server.test.identity.0.principal_id
}
resource "azurerm_mssql_server_extended_auditing_policy" "test" {
  server_id        = azurerm_mssql_server.test.id
  storage_endpoint = azurerm_storage_account.test.primary_blob_endpoint
  depends_on = [
    azurerm_role_assignment.test,
  ]
}
```
Please try if it works.

**Comment by eugeneromero** [comment]
@yupwei68 thank you for your reply! I tested your suggestion and indeed, it seems the trick is adding the `"Storage Blob Data Contributor"` permission. Thank you for pointing me in the right direction!

For anyone coming here looking for this, switching from the "inline" auditing policy to the "azurerm_mssql_server_extended_auditing_policy" resource was not straightforward. I had to do the following steps in order:

1) Go into the Azure Portal and turn off auditing in the SQL server
2) Remove the inline block and add the two new blocks
3) Ran a tf plan with refresh, `terraform plan -refresh=true -out=terraform.tfplan`
4) Applied the changes

Otherwise, Terraform does not see the removal of the inline block, and keeps failing to create the new resource because 
```
A resource with the ID "/subscriptions/[...]/extendedAuditingSettings/Default" already exists - to be managed via Terraform this resource needs to be imported into the State
```
but the resource cannot be moved or imported, because it does not have its own terraform id. Hope this helps!

I will close this issue now. Thanks again!
